compiler/rustc_ty_utils/src/instance.rs

```diff
@@ -2,14 +2,123 @@ use rustc_errors::ErrorReported;
 
+// FIXME(#86795): `BoundVarsCollector` here should **NOT** be used
+// outside of `resolve_associated_item`. It's just to address #64494,
+// #83765, and #85848 which are creating bound types/regions that lose
+// their `Binder` *unintentionally*.
+// It's ideal to remove `BoundVarsCollector` and just use
+// `ty::Binder::*` methods but we use this stopgap until we figure out
+// the "real" fix.
+struct BoundVarsCollector<'tcx> {
+    binder_index: ty::DebruijnIndex,
+    vars: BTreeMap<u32, ty::BoundVariableKind>,
+    // We may encounter the same variable at different levels of binding, so
+    // this can't just be `Ty`
+    visited: SsoHashSet<(ty::DebruijnIndex, Ty<'tcx>)>,
+}
+
+impl<'tcx> BoundVarsCollector<'tcx> {
+    fn new() -> Self {
+        BoundVarsCollector {
+            binder_index: ty::INNERMOST,
+            vars: BTreeMap::new(),
+            visited: SsoHashSet::default(),
+        }
+    }
+
+    fn into_vars(self, tcx: TyCtxt<'tcx>) -> &'tcx ty::List<ty::BoundVariableKind> {
+        let max = self.vars.iter().map(|(k, _)| *k).max().unwrap_or(0);
+        for i in 0..max {
+            if let None = self.vars.get(&i) {
+                panic!("Unknown variable: {:?}", i);
+            }
+        }
+
+        tcx.mk_bound_variable_kinds(self.vars.into_iter().map(|(_, v)| v))
+    }
+}
+
+impl<'tcx> TypeVisitor<'tcx> for BoundVarsCollector<'tcx> {
+    type BreakTy = ();
+
+    fn visit_binder<T: TypeFoldable<'tcx>>(
+        &mut self,
+        t: &Binder<'tcx, T>,
+    ) -> ControlFlow<Self::BreakTy> {
+        self.binder_index.shift_in(1);
+        let result = t.super_visit_with(self);
+        self.binder_index.shift_out(1);
+        result
+    }
+
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+        if t.outer_exclusive_binder() < self.binder_index
+            || !self.visited.insert((self.binder_index, t))
+        {
+            return ControlFlow::CONTINUE;
+        }
+        use std::collections::btree_map::Entry;
+        match *t.kind() {
+            ty::Bound(debruijn, bound_ty) if debruijn == self.binder_index => {
+                match self.vars.entry(bound_ty.var.as_u32()) {
+                    Entry::Vacant(entry) => {
+                        entry.insert(ty::BoundVariableKind::Ty(bound_ty.kind));
+                    }
+                    Entry::Occupied(entry) => match entry.get() {
+                        ty::BoundVariableKind::Ty(_) => {}
+                        _ => bug!("Conflicting bound vars"),
+                    },
+                }
+            }
+
+            _ => (),
+        };
+
+        t.super_visit_with(self)
+    }
+
+    fn visit_region(&mut self, r: ty::Region<'tcx>) -> ControlFlow<Self::BreakTy> {
+        use std::collections::btree_map::Entry;
+        match r {
+            ty::ReLateBound(index, br) if *index == self.binder_index => match br.kind {
+                ty::BrNamed(_def_id, _name) => {
+                    // FIXME
```

Comment:
Oh oops. I don't remember if later versions of this added this or if I begin to track state before it was needed.

Given that the problem here is still unclear, I wonder if it's worth adding back this code or not (and for BrEnv). At the very least the FIXME should be changed to an explicit comment.

Also...it just occured to me that we now have a `var` field available to us. So we can rewrite this in a smarter way. Or, at least verify that the variable is in the same place we would expect it!

I can either look back through the history to see if there's a more up to date version, or maybe you can.

Reply:
> I can either look back through the history to see if there's a more up to date version, or maybe you can.

So after the above commit, https://github.com/rust-lang/rust/pull/76814/commits/6d5efa9f040d2638318f9e8b96eab718ef664e3c changed it to the current style.

> Also...it just occured to me that we now have a var field available to us. So we can rewrite this in a smarter way. Or, at least verify that the variable is in the same place we would expect it!

Hmm, I haven't understood it, could you elaborate? Something like this?

```rust
ty::BrNamed(def_id, _name) => match self.vars.get(&br.var.as_u32()) {
                    Some(_) => {}
                    None => bug!("couldn't find!"),
                }
```

Reply:
Oops, this is what I get for reviewing from mobile early. I think we can just do
```rust
match r {
  ty::ReLateBound(index, br) if *index == self.binder_index => {
    match self.vars.entry(br.var.as_u32()) {
      ... (insert of check same kind)
    }
  }
}
```

Reply:
Thanks! Actually, I overlooked how we implemented `visit_ty` above. Now I think I addressed your comment.